File: server/models/order_item.ts
import { DataTypes, Model, Optional, Sequelize } from 'sequelize';

interface OrderItemAttributes {
  id: string;
  order_id: string;
  product_id?: string;
  product_name: string;
  quantity: number;
  price_per_unit: number;
  affiliate_id?: string;
}

export interface OrderItemCreationAttributes extends Optional<OrderItemAttributes, 'id'> {}

export class OrderItem extends Model<OrderItemAttributes, OrderItemCreationAttributes> implements OrderItemAttributes {
  public id!: string;
  public order_id!: string;
  public product_id?: string;
  public product_name!: string;
  public quantity!: number;
  public price_per_unit!: number;
  public affiliate_id?: string;

  // timestamps
  public readonly created_at!: Date;
  public readonly updated_at!: Date;

  public static associate: (models: any) => void;
}

export function initOrderItem(sequelize: Sequelize) {
  OrderItem.init(
    {
      id: {
        type: DataTypes.STRING(36), // UUID
        primaryKey: true,
        defaultValue: DataTypes.UUIDV4,
      },
      order_id: {
        type: DataTypes.STRING(36),
        allowNull: false,
        references: { model: 'orders', key: 'id' },
        onDelete: 'CASCADE',
      },
      product_id: {
        type: DataTypes.STRING(36),
        references: { model: 'products', key: 'id' },
        onDelete: 'SET NULL',
      },
      product_name: { type: DataTypes.STRING, allowNull: false },
      quantity: { type: DataTypes.INTEGER, allowNull: false },
      price_per_unit: { type: DataTypes.DECIMAL(10, 2), allowNull: false },
      affiliate_id: {
        type: DataTypes.STRING(36),
        references: { model: 'affiliate_profiles', key: 'id' },
        onDelete: 'SET NULL',
      },
    },
    {
      sequelize,
      modelName: 'OrderItem',
      tableName: 'order_items',
      timestamps: true, // Let Sequelize manage timestamps
      underscored: true // Use snake_case for all field names
    }
  );

  OrderItem.associate = function(models) {
    OrderItem.belongsTo(models.Order, {
      foreignKey: 'order_id',
      as: 'order',
    });
    OrderItem.belongsTo(models.Product, {
      foreignKey: 'product_id',
      as: 'product',
    });
    OrderItem.belongsTo(models.AffiliateProfile, {
      foreignKey: 'affiliate_id',
      as: 'affiliate',
    });
  };

  return OrderItem;
}

File: server/models/order_item.py
import uuid
from sqlalchemy.sql import func
from sqlalchemy import Numeric, String, Integer, DateTime, ForeignKey
from . import db

class OrderItem(db.Model):
    __tablename__ = 'order_items'
    
    id = db.Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    order_id = db.Column(String(36), ForeignKey('orders.id', ondelete='CASCADE'), nullable=False)
    product_id = db.Column(String(36), ForeignKey('products.id', ondelete='SET NULL'), nullable=True)
    product_name = db.Column(String(255), nullable=False)
    quantity = db.Column(Integer, nullable=False, default=1)
    price_per_unit = db.Column(Numeric(10, 2), nullable=False)
    affiliate_id = db.Column(String(36), ForeignKey('affiliate_profiles.id', ondelete='SET NULL'), nullable=True)
    created_at = db.Column(DateTime, default=func.now(), nullable=False)
    updated_at = db.Column(DateTime, default=func.now(), onupdate=func.now(), nullable=False)

    # Relationships
    product = db.relationship('Product', back_populates='order_items')
    
    def to_dict(self):
        """Convert order item object to dictionary"""
        return {
            'id': self.id,
            'order_id': self.order_id,
            'product_id': self.product_id,
            'product_name': self.product_name,
            'quantity': self.quantity,
            'price_per_unit': float(self.price_per_unit) if self.price_per_unit is not None else None,
            'affiliate_id': self.affiliate_id,
            'created_at': self.created_at.isoformat() if self.created_at else None,
            'updated_at': self.updated_at.isoformat() if self.updated_at else None,
            'product': self.product.to_dict() if self.product else None
        }
